File: Medium/2026-02-13-3714-Longest-Balanced-Substring-II/solution.py
class Solution:
    def longestBalancedSubstring(self, s: str) -> int:
        max_length = 0
        n = len(s)

        for i in range(n):
            count = {}
            for j in range(i, n):
                count[s[j]] = count.get(s[j], 0) + 1
                if len(count) > 3:
                    break
                if len(count) == 1:
                    continue
                if len(count) == 2:
                    if len(set(count.values())) == 1:
                        max_length = max(max_length, j - i + 1)
                elif len(count) == 3:
                    if len(set(count.values())) == 1:
                        max_length = max(max_length, j - i + 1)

        return max_length

File: Medium/2026-02-13-3714-Longest-Balanced-Substring-II/solution.java
class Solution {
    public int longestBalancedSubstring(String s) {
        int n = s.length();
        int maxLength = 0;

        for (int targetCount = 1; targetCount <= n; targetCount++) {
            int[] count = new int[3]; // count for 'a', 'b', 'c'
            int distinctChars = 0;

            for (int i = 0; i < n; i++) {
                int index = s.charAt(i) - 'a';
                count[index]++;
                if (count[index] == targetCount) {
                    distinctChars++;
                }
                if (distinctChars == 3) {
                    maxLength = Math.max(maxLength, i + 1);
                }
                if (count[index] > targetCount) {
                    while (count[index] > targetCount) {
                        count[s.charAt(i - (i - (i - 1) + 1)) - 'a']--;
                        if (count[s.charAt(i - (i - 1) + 1) - 'a'] == targetCount - 1) {
                            distinctChars--;
                        }
                        i--;
                    }
                }
            }
        }

        return maxLength;
    }
}